mino/minogrpc/overlay.go

```diff
@@ -177,18 +184,30 @@ func (o overlayService) Stream(stream Overlay_StreamServer) error {
 		newCtx := stream.Context()
 		newCtx = metadata.NewOutgoingContext(newCtx, header)
 
-		clientStream, err := cl.Stream(newCtx)
+		cs, err := cl.Stream(newCtx)
 		if err != nil {
 			err = xerrors.Errorf("failed to get stream for client '%s': %v",
 				addr.String(), err)
 			fabric.Logger.Err(err).Send()
 			return err
 		}
 
-		sender.participants[addr.String()] = clientStream
+		clientStream := newSafeOverlayStream(cs)
+
+		sender.participants.Store(addr.String(), clientStream)
 
 		// Sending the routing info as first messages to our children
-		clientStream.Send(&OverlayMsg{Message: overlayMsg.Message})
+		clientStream.Send(&Envelope{
+			From:         o.addr.String(),
+			PhysicalFrom: o.addr.String(),
+			To:           []string{addr.String()},
+			Message:      envelope.Message,
+		})
+
+		// TODO: think how we can make all the node set the routing then start
```

Comment:
It should be impossible to receive a message before the routing message, because the relay will open the connection and then directly send the routing message before relaying anything.